From 8082f0e98e9c54f6a835f0220d48ce6615f2f002 Mon Sep 17 00:00:00 2001
From: Hauke Mehrtens <hauke.mehrtens@intel.com>
Date: Wed, 5 Sep 2018 17:34:00 +0200
Subject: [PATCH] PONRTSYS-1977: gpio: intel-sso: Deactivate not used SOUT
 groups

This will make the driver actively deactivate the SOUT groups which are not
used and not only activate the needed groups.

The reset value of this register is 0, but the U-Boot can also use the SSO
registers and could activate all groups.
This is needed to restrict the SSO to 2 groups on the PRX321 SFU board.
---
 drivers/gpio/gpio-intel-sso.c | 17 +++++++++++------
 1 file changed, 11 insertions(+), 6 deletions(-)

diff --git a/drivers/gpio/gpio-intel-sso.c b/drivers/gpio/gpio-intel-sso.c
index 9b3c19b7898b..5f19fa31793b 100644
--- a/drivers/gpio/gpio-intel-sso.c
+++ b/drivers/gpio/gpio-intel-sso.c
@@ -163,19 +163,24 @@ static int sso_gpio_gc_init(struct sso_gpio_priv *priv,
 static int sso_gpio_hw_init(struct sso_gpio_priv *priv)
 {
 	int i;
+	int err;
+	u32 activate;
 
 	/* Clear all duty cycles */
 	for (i = 0; i < priv->pins; i++) {
-		if (sso_gpio_writel(priv->mmap, DUTY_CYCLE(i), 0))
-			return -ENOTSUPP;
+		err = sso_gpio_writel(priv->mmap, DUTY_CYCLE(i), 0);
+		if (err)
+			return err;
 	}
 
 	/* 4 groups for total 32 pins */
 	for (i = 1; i <= MAX_GROUP_NUM; i++) {
-		if (i * PINS_PER_GROUP <= priv->pins ||
-		    priv->pins > (i - 1) * PINS_PER_GROUP)
-			if (sso_gpio_update_bit(priv->mmap, SSO_CON1, i - 1, 1))
-				return -ENOTSUPP;
+		activate = !!(i * PINS_PER_GROUP <= priv->pins ||
+			      priv->pins > (i - 1) * PINS_PER_GROUP);
+		err = sso_gpio_update_bit(priv->mmap, SSO_CON1, i - 1,
+					  activate);
+		if (err)
+			return err;
 	}
 
 	/* NO HW directly controlled pin by default */
